### DISCUSSION TITLE
Cookbook ideas

### DISCUSSION THREAD

**Discussion by jukent** [ProjectPythia/projectpythia.github.io#248] [Ideas]
Ideas from today's ESDS forum. Please comment yours!

**Comment by jukent** [comment]
- How best to use Binder with large data that is stored somewhere else? (`intake` most likely!)

- Best practices for running Jupyter on Supercomputers (how to make sure you shut down your server, e.g.)

- Environment management best practices and troubleshooting techniques to solve certain problems (downgrade a package etc -- will need more examples of these problems - should we show how GitHub actions does this for us?)

- Making a Python package

**Comment by samsrabin** [comment]
We have some cookbook-type notebooks for CTSM at the [ctsm_python_gallery](https://github.com/NCAR/ctsm_python_gallery) repo. At least some of these are based on a set of utilities ([utils.py](https://github.com/NCAR/ctsm_python_gallery/blob/master/ctsm_py/utils.py)) for doing various things with `xarray` including:

- Convert DataArray units
- Clean up DataArray unit strings
- Get a weighted annual mean from monthly outputs
- Set Dataset `time` values to midpoint of `time_bounds` values
- Get time as year + fraction of year
- Add a cyclic coordinate to a DataArray
- Convert a latitude axis that's -180 to 180 around the international date line to one that's 0 to 360 around the prime meridian
- As the above, but vice-versa
- Read multiple netCDF files to in-memory Dataset, optionally removing unneeded variables and/or slices.
- Grid a variable from a netCDF in the collapsed-grid format (with `patches1d_ixy` etc.)

Would be great to see if y'all have any optimizations or generalizations for those functions as well!

**Reply by clyne** [reply]
@samsrabin while we're' still trying to figure out what the definition of a 'cookbook' is, some of the useful things that you describe sound more like Xarray 'operators':  functions that consume and possibly produce new Xarray objects. If you think some of your utilities might be of general use to the geoscience communities, please consider contributing them to [GeoCAT](https://geocat.ucar.edu/). 

CC: @erogluorhan

**Reply by samsrabin** [reply]
Yes, everything in `utils.py` is a function, but there are some notebooks in that repo that I think I'd call "cookbooks." Or at least, they're examples that could be translated into cookbooks.

I didn't know about GeoCAT, looks useful. Thanks!

**Reply by erogluorhan** [reply]
Both GeoCAT and CTSM's utils even have some overlapping functions (e.g. annual mean, cyclic longitudes, etc.) already. Your cookbook notebooks look great, providing examples of how to use those functions, @samsrabin ! I second @clyne here, in addition to having those functions showcased in your notebooks, if you'd like to deliver them to the existing GeoCAT user base, please consider contributing them (we are happy to credit any individuals and groups). 

Additionally, I'd like to point you out to our GeoCAT-examples, for example this [calendar average example](https://geocat-examples.readthedocs.io/en/latest/gallery-geocat-comp/matplotlib/calendar_average_example.html#sphx-glr-gallery-geocat-comp-matplotlib-calendar-average-example-py), which demonstrates examples of analysis functions in use, in hopes it might be helpful.

Best,

**Comment by clyne** [comment]
I've created a shared draft [document](https://docs.google.com/document/d/15Vn6fwtRyBC8LtffJWGPM4FUB31QepMGOtolC9hWTRw/edit?usp=sharing) for planning the hackathon. Please review and edit as you see fit.

**Comment by r-ford** [comment]
I'm working on a cookbook that starts with some of the Pangeo Gallery material. It will probably include using
- [Intake-ESM](https://intake-esm.readthedocs.io/en/latest/index.html) ([Pangeo Gallery notebook](http://gallery.pangeo.io/repos/pangeo-gallery/cmip6/intake_ESM_example.html))
- [CMIP6 Preprocessing](https://cmip6-preprocessing.readthedocs.io/en/latest/index.html)
- [xarrayutils](https://xarrayutils.readthedocs.io/en/latest/index.html)
- [xESMF](https://xesmf.readthedocs.io/en/latest/)

The goal is to start with the CMIP6 data on Pangeo's Google Storage and end with a certain plot of the multimodel mean ocean heat uptake ([specifically Fig. 1b in this paper](https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2019AV000130)).

For another cookbook, I think it would be cool to do some sort of EOF analysis with the [eofs package](https://ajdawson.github.io/eofs/latest/), probably also using that CMIP6 data.

**Reply by erogluorhan** [reply]
You may find checking the GeoCAT's EOF analysis functions that use the eofs package under the hood also useful.

**Comment by jukent** [comment]
From Teagan King - scripts for interpolating data from terrain following model levels to specific pressure levels

**Reply by ktyle** [reply]
@kmpaul is this vertical interpolation capability going to be in the initial release of[ xwrf](https://xwrf.readthedocs.io/)? Cookbooks related to WRF would be quite popular ...

**Reply by kmpaul** [reply]
It doesn't look like it. The capability is getting fixed in xgcm, and xwrf is going to use that, but we are waiting for xgcm.
